iterate through blobs before checking prefixes

Lee-W opened the pull request:
After reading https://cloud.google.com/storage/docs/json_api/v1/objects/list, https://github.com/googleapis/python-storage/blob/v2.14.0/google/cloud/storage/client.py#L1143-L1145C35 and https://github.com/apache/airflow/blob/providers-google/10.13.0rc2/airflow/providers/google/cloud/hooks/gcs.py#L732, I suspect we might not use the `delimiter` the right way. But it's going to be deprecated. So it might be better for us to keep the original behavior.

According to https://github.com/googleapis/python-storage/blob/v2.14.0/google/cloud/storage/client.py#L1213-L1217, the prefixes are not returned until the `blobs` are consumed. Thus, to keep the behavior of checking whether `prefixes` exists and decide the extended content, we'll need to consume `blobs` first

Related: https://github.com/apache/airflow/pull/36130


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

atrbgithub commented on airflow/providers/google/cloud/hooks/gcs.py:
```diff
@@ -933,16 +934,17 @@ def list_by_timespan(
                 )
-                list(blobs)
+
+            blob_names = [
+                blob.name
+                for blob in blobs
+                if timespan_start <= blob.updated.replace(tzinfo=timezone.utc) < timespan_end
+            ]
```
This PR looks great and returns the provider to it's original behaviour, it's working fine for me when testing locally.

It's worth noting though that when a delimiter is passed in, the filter here has no effect. It will simply return all files. The filtering only works when `match_glob` is used. 

This is how the provider has been working up until this point however so this can be addressed later, if ever as delimiter is deprecated.

Lee-W committed:
fix(providers/google): iterate through blobs before checking prefixes

According to https://github.com/googleapis/python-storage/blob/v2.14.0/google/cloud/storage/client.py#L1213-L1217, the prefixes are not returned until the blobs are consumed

sunank200 reviewed (approved):
LGTM

phanikumv reviewed (commented):
Please add tests

Lee-W committed:
test(providers/google): add test cases to check gcs.list result
